### ISSUE TITLE
For a class c1 implementing module c1.py, robotframework creates instances of other classes (c2)/modules imported inside the former class c1

### ISSUE THREAD

**Issue by sudarsesh** [robotframework/robotframework#3124]
This is the following directory structure that I have,

```bash
.
|-- par
|   |-- __init__.py
|   |-- par2
|   |   |-- __init__.py
|   |   |-- c1.py
|   |   |-- c2.py
|   |   `-- del1
|   |       |-- __init__.py
|   |       |-- d1.py
|   |       |-- del2
|   |       |   |-- __init__.py
|   |       |   |-- d2.py
|   |       `-- test.py
|   |-- print-dir.py
`-- run.robot
```

Contents of run.robot

```robotframework
*Settings*

Library   par/par2/c1.py
Library   c2 

*Test cases*
T1
    Run keyword and continue on failure  print c1
    Run keyword and continue on failure  print c2
    ${c1}=  Run keyword and continue on failure  Get library instance   c1
    ${c2}=  Run keyword and continue on failure  Get library instance   c2
```

Contents of ./par/par2/c1.py

```python
#from par.par2.c2 import c2
from c2 import c2
from robot.api import logger
from pprint import pprint as pp

class c1(object):
    pass

    def print_c1(self):
        logger.info("globals()\n")
        pp(globals())
```

Contents of ./par/par2/c2.py

```python
class c2(object):
    pass

    def print_c2(self):
        logger.info("globals()\n")
        pp(globals())
```
**Scenario-1:**

Both class c1 and c2 are classes implementing modules. In the module c1.py, I have defined the class c1.py and have also imported the class c2 from the module c2.

In the module c2.py, I have only defined the class c2.

In the robot script, I have imported the Library c1 by specifying the path to c1.py (assuming the parent folder containing the script is added to the python path). I have not specified the path to c2.py but only sourced the library as c2 (c2 will be the name ref for that module).

To my surprise, the script ran successfully and I am able to get both the library instances in spite of not specifying the path to c2.py in the script while importing that library. 

PFA for the logs.zip

However the same doesnt work when I import c2 using absolute path instead of relative path (since both c1 and c2 exists in the same folder)

**Scenario-2:**

```python
from par.par2.c2 import c2
#from c2 import c2
```
Now when I run the script, the script fails at the keyword,

```robotframework
${c2}=  Run keyword and continue on failure  Get library instance   c2
```
PFA for the logs.zip

Upon inspecting the globals info in c1.py, I could see that RobotFramework only considers the module name and not the local name mapped to the module in c1.py,

```python
#For scenario-1
'c1': <class 'c1.c1'>,
'c2': <class 'c2.c2'>

#For scenario-2
'c1': <class 'c1.c1'>,
'c2': <class 'par.par2.c2.c2'>
```

It stores the module object for c2 in scenario-1 under the module name 'c2' and as 'par.par2.c2' in scenario-2. So when we want to source this module in the robot script, we have to import the library c2 with the module name stored in globals 'par.par2.c2' and not use the local name 'c2'.

**Scenario-3:**

```robotframework
#Changed the Library name to import by specifying the module name associated with c2
#Library      c2
Library   par.par2.c2     WITH NAME   c2
```

PFA for the logs.zip

So there are basically 2 surprises which I couldnt find it in the robotframework user guide. Please correct me if I have missed it out in the user guide.

1. RobotFramework not only creates instances of classes implementing modules, but also for instances of other classes/modules imported inside the former such classes (here former is c1.py)
2. The name associated with such instances is the name of the module and not the local name associated with such modules. Hence to source such instances inside the robotscript, we must use the module path and not the local name.
3. The instances of others classes imported inside a particular module having the same name as its class, are available in the robotframework for name-binding inside the script using the **Library** setting or **ImportLibrary** keyword   
I think this is an important information that can be specified in the robotframework user guide.
[logs.zip](https://github.com/robotframework/robotframework/files/2977075/logs.zip)

**Comment by sudarsesh** [comment]
Hi @pekkaklarck ,
Could you please help?

**Comment by pekkaklarck** [comment]
I got somewhat lost trying to read the description. Could you summarize what's the actual problem you have?

**Comment by sudarsesh** [comment]
So there are basically 2 surprises which I couldnt find it in the robotframework user guide. Please correct me if I have missed it out in the user guide.

RobotFramework not only creates instances of classes implementing modules, but also for instances of other classes/modules imported inside the former such classes (here former is c1.py)
The name associated with such instances is the name of the module and not the local name associated with such modules. Hence to source such instances inside the robotscript, we must use the module path and not the local name.
The instances of others classes imported inside a particular module having the same name as its class, are available in the robotframework for name-binding inside the script using the Library setting or ImportLibrary keyword
I think this is an important information that can be specified in the robotframework user guide.

**Comment by pekkaklarck** [comment]
I'm still pretty confused. First of all, classes don't implement modules. Secondly, I'm pretty certain Robot Framework itself only creates instances of classes imported as libraries.

Anyway, I think the problem you have faced is related to Python storing imported modules in `sys.modules` to make importing them again faster. My guess is that when in your first example you for `from c2 import c2` in  `c1.py`, that makes `c2` module available in `sys.modules` and thus later `Library     c2` works. Without that `from c2 import c2` the import would fail because `c2` module isn't found from `PYTHONPATH` (i.e. `sys.path`). That happens also when you do `from par.par2.c2 import c2` because then `par` module, not `c2`, is added to `sys.modules`.

In general your setup loops pretty complicated. I don't understand why you'd need such a nested module structure and why you'd want to import some (sub)modules using path (`Library    par/par2/c1.py`) and others using just a library name (`Library    c2`) that requires the module to be in `PYTHONPATH`. Just having your `c1.py` and `c2.py` in a directory added to `PYTHONPATH` would be so much simpler.

Anyway, if you got concrete ideas how to enhance the User Guide let me know. I don't think explaining how Python module importing and caching works belongs there, though, but possibly a link to some external Python documentation could be added.
